R/data.table.R

```diff
@@ -3086,3 +3110,153 @@ isReallyReal = function(x) {
+      } 
+    } else {
+
+      i_j_subset = substitute(1:min(n, .N), list(n = abs(n)))
+      if (n < 0) i_order = substitute(order(wt), list(wt = substitute(wt))) else i_order = substitute(order(-wt), list(wt = substitute(wt)))
+      
+      if (missing(by)){
+        l_args[['i']] = substitute(order_i[i_subset], list(order_i = i_order, i_subset = i_j_subset))
+        l_args[['f_order']] = TRUE
+      } else {
+        l_args[['i']] = i_order
+        l_args[['j']] = substitute(.I[j_subset], list(j_subset = i_j_subset))
+      }
+    }
+  }
+
+  l_args
+}
+
+filter.at = function(cols, logic, by, all.vars = TRUE){
+  ## helper that takes takes filter.at() from the isub and massages it to produce indices for isub
+  
+  #' @param cols required; can accept all values that .SDcols accepts
+  #' @param logic required; a function or unquoted text that results in logic evaluation. The
+  #'              unquoted text must include `x` as an argument.
+  #'              If the argument contains `which.min` or `which.max`, intermediate results will
+  #'              be combined with `union` instead of `&` or `|`. 
+  #' @param by optional; accepts anything that the normal by argument accepts.
+  #' @param all.vars logical. Except as mentioned in `logic`, `all.vars = T` will reduce
+  #'                  results using `&` and `all.vars = F` will reduce results using `|`
+  #'                  
+  #' @return List. The number of list elements depends on the arguments above. 
+  #' 
+  #' `filter.at(cols = T, logic, all.vars = TRUE)` returns an unnamed list that needs to
+  #'      be evaulated in loop. This is to allow for some optimization of `dt[a & b & c]`.
+  #'      Argument can also be `cols = c('V1','V2')` to get the same  return as `cols = T`.
+  #' 
+  #' All other combinations return a named list to be evaluated as `do.call(`[data.table`, l_args)`
+  
+  cols = substitute(cols)
+  logic = substitute(logic)
+  
+  # for anonymous functions vs. functions in the environment
+  is_fx = any(as.character(logic) == 'x') || length(as.character(logic)) > 1
+
+  # Predicates wtih rle and cumsum do not work with chaining i.e. 
+  #
+  # dt[rleid(x1) < 5 & rleid(x2) < 5] does not always equal
+  # dt[rleid(x1) < 5][rleid(x2) < 5]
+  #
+  # May want to make this an arg that defaults to FALSE or based on data.table optimization arg
+
+  quickFilter = !grepl('rle|cumsum|min|max|sum', deparse(logic))
+  which.mFilter = grepl('which.m', deparse(logic))
+
+  if (missing(by) && all.vars && quickFilter && !which.mFilter && is_fx && !grepl("patterns|:", deparse(cols))){
+    
+    #optimization. For logic of x == 5 and cols = c('a','b','c') the result of this would be
+    #
+    # ind <- which(dt[['a']] == 5)
+    # for(col in c('b','c')){
+    #   ind <- ind[dt[['col']][ind] == 5]
+    # }
+    
+    #use `<-` within substitution; changing to `=` caused errors
```

Comment:
Because by using `=` you are trying to provide named argument to `substitute` function (see `args(substitute)` and not assign a new variable. This is a good example where we precisely we need to use `<-`